### PR TITLE
26-as-a-user-i-want-to-validate-if-metadata-is-hip-412-compliant

### LINKED ISSUE #26: As a User I want to validate if metadata is HIP-412 compliant
**Acceptance criteria:**
- the SDK can validate local directory with JSON files
- the SDK can validate single, local JSON file
- the SDK can validate all NFTs JSON metadata from an on-chain collection
  - Download collection metadata to a local directory
  - validate locally
- the SDK can validate single, on-chain NFT metadata
- the SDK should return:
  - a path to a file that contains validation errors
    - In case of non-compliant files/nfts, it should save errors for each non-compliant file/nft to a directory.
  - a success flag that should be true if there are no validation errors
- the validation logic should try to use code from hedera-nft-utilities / mintbox

file naming convention should be `operationType-date-hour.txt`

example: `HIP412-validation-290124-152730.txt`

### PR DESCRIPTION
**Pull request by michalrozekariane**
**Description**:

**Related issue(s)**:

Fixes #26 

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: src/utils/services/Hip412Validator.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    const jsonFiles = fs
+      .readdirSync(directoryPath)
+      // .gitkeep file is needed inside empty-json-directory to keep it in the repository that's why we filter it out
+      .filter((file) => file !== '.gitkeep')
+      .sort((a, b) => {
+        const numA = parseInt(a.match(/\d+/)?.[0] ?? '0', 10);
+        const numB = parseInt(b.match(/\d+/)?.[0] ?? '0', 10);
+        return numA - numB;
+      });
```
Comment (BartoszSolkaArianelabs) [inline src/utils/services/Hip412Validator.ts:115]: why is it required to sort the files?

File: test/unit/validateSingleMetadataObject.test.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  });
+
+  it('should return an error for an object with an invalid attributes structure', () => {
+    const validationResult = validate(METADATA_OBJECT_WITH_INVALID_ATTRIBUTES_STRUCTURE);
+    expect(validationResult.isValid).toBe(false);
+    expect(validationResult.errors.general).toEqual([
+      dictionary.validation.requiredAttributeFieldMissing,
+    ]);
+  });
+});
```
Comment (BartoszSolkaArianelabs) [inline test/unit/validateSingleMetadataObject.test.ts:89]: please add a test for validating object with multiple errors

**Review by BartoszSolkaArianelabs — CHANGES_REQUESTED** [review]

**Review by BartoszSolkaArianelabs — CHANGES_REQUESTED** [review]

File: src/utils/services/Hip412Validator.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  }
+
+  static validateLocalDirectory(directoryPath: string): DirectoryValidationResult {
+    const errors: MetadataError[] = [];
+
+    const jsonFiles = fs
+      .readdirSync(directoryPath)
+      // .gitkeep file is needed inside empty-json-directory to keep it in the repository that's why we filter it out
+      .filter((file) => file !== '.gitkeep')
```
[the remark this answers is not in the document]
Comment (michalrozekariane) [reply src/utils/services/Hip412Validator.ts:112]: done

File: test/unit/validateSingleMetadataObject.test.ts (hunk 2 above)
Comment (michalrozekariane) [reply test/unit/validateSingleMetadataObject.test.ts:96]: done

File: src/api/mirrorNode.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+          error.response.status
+        );
+      } else {
+        console.log(dictionary.errors.unknownErrorWhileFetching(serialNumber));
+      }
+    }
+
+    return {
+      serialNumber,
+      error: errorMessage,
```
Comment (se7enarianelabs) [inline src/api/mirrorNode.ts:89]: Wouldn't it be good if it would also return a status if successful or failed?

File: src/types/nfts.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+export interface NFTDetails {
+  account_id: string;
+  created_timestamp: string;
+  deleted: boolean;
+  metadata: string;
+  modified_timestamp: string;
+  serial_number: number;
+  token_id: string;
+  meta?: NFTMetadata;
```
Comment (se7enarianelabs) [inline src/types/nfts.ts:11]: Why meta and not metadata?

File: src/utils/helpers/nftMetadataDecoder.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import { dictionary } from '../constants/dictionary';
+import { errorToMessage } from './errorToMessage';
+
+export const nftMetadataDecoder = (
+  nfts: NFTDetails | NFTDetails[],
+  ipfsGateway?: string
+): DecodedMetadata[] => {
+  const nftsArray = Array.isArray(nfts) ? nfts : [nfts];
+  const decodedMetadataArray: DecodedMetadata[] = nftsArray.map((nft: NFTDetails) => {
+    let decodedNFTMetadata: string;
```
Comment (se7enarianelabs) [inline src/utils/helpers/nftMetadataDecoder.ts:11]: Initialize it with a default value like "", it's a good practice

File: src/utils/helpers/nftMetadataDecoder.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import { NFTDetails, DecodedMetadata } from '../../types/nfts';
+import { dictionary } from '../constants/dictionary';
+import { errorToMessage } from './errorToMessage';
+
+export const nftMetadataDecoder = (
```
Comment (se7enarianelabs) [inline src/utils/helpers/nftMetadataDecoder.ts:5]: rename function to a URI decoder, as it does not decode the metadata,
I thought that it downloads/decodes the final metadata from the NFTs uri

File: src/utils/services/Hip412Validator.ts [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+          Hip412MetadataCSVSchema,
+          metadataObject,
+          validationMetadataErrorOptions
+        );
+      } catch (e) {
+        errors.push(
+          dictionary.validation.arrayOfObjectsValidationError(
+            filePath || `object ${index + 1}`,
+            errorToMessage(
+              errorToMessage(e) === 'Required' ? dictionary.validation.requiredFieldMissing : e
```
Comment (se7enarianelabs) [inline src/utils/services/Hip412Validator.ts:83]: extract `'Required'` to a const

File: src/utils/services/Hip412Validator.ts [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+    for (const [index, metadataObject] of metadataObjects.entries()) {
+      try {
+        validateObjectWithSchema(
+          Hip412MetadataCSVSchema,
+          metadataObject,
+          validationMetadataErrorOptions
+        );
+      } catch (e) {
+        errors.push(
```
Comment (se7enarianelabs) [inline src/utils/services/Hip412Validator.ts:79]: in validateSingleMetadataObject in the catch statement you check if the err is an instanceof ValidationError.
Here you don't do it, why?

File: src/utils/services/Hip412Validator.ts [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+
+    try {
+      validateObjectWithSchema(Hip412MetadataSchema, object, validationMetadataErrorOptions);
+    } catch (err) {
+      if (err instanceof ValidationError) {
+        errors.push(...err.errors);
+      } else {
+        errors.push(errorToMessage(err));
+      }
+    }
```
Comment (se7enarianelabs) [inline src/utils/services/Hip412Validator.ts:57]: You could extract this fragment and use it in validateArrayOfObjects as well

File: src/utils/services/Hip412Validator.ts [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+
+    const filesForValidation = fs
+      .readdirSync(directoryPath)
+      // .gitkeep file is needed inside empty-json-directory to keep it in the repository that's why we filter it out
+      .filter((file) => (file.endsWith('.json') || file.endsWith('.txt')) && file !== '.gitkeep')
+      .sort((a, b) => {
+        const numA = parseInt(a.match(/\d+/)?.[0] ?? '0', 10);
+        const numB = parseInt(b.match(/\d+/)?.[0] ?? '0', 10);
+        return numA - numB;
+      });
```
Comment (se7enarianelabs) [inline src/utils/services/Hip412Validator.ts:116]: Why do we need this?

**Review by se7enarianelabs — CHANGES_REQUESTED** [review]

File: src/api/mirrorNode.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+};
+
+export async function getMetaDataFromMirrorNode(
+  network: NetworkName,
+  nftId: NftId,
+  mirrorNodeUrl?: string
+): Promise<string> {
+  const url = mirrorNodeUrl || getMirrorNodeUrlForNetwork(network);
+  return axios
+    .get(`${url}/tokens/${nftId.tokenId.toString()}/nfts/${nftId.serial.toString()}`)
```
[the remark this answers is not in the document]
Comment (michalrozekariane) [reply src/api/mirrorNode.ts:20]: done

File: src/api/mirrorNode.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+      return atob(response.data.metadata);
+    });
+}
+
+export async function getNFTsFromToken(
+  network: NetworkName,
+  tokenId: string
+): Promise<NFTDetails[]> {
+  const baseUrl = getMirrorNodeUrlForNetwork(network);
+  let nextLink: string = `${baseUrl}/tokens/${tokenId}/nfts`;
```
[the remark this answers is not in the document]
Comment (michalrozekariane) [reply src/api/mirrorNode.ts:32]: done

File: src/utils/helpers/nftMetadataDecoder.ts (hunk 7 above)
Comment (michalrozekariane) [reply src/utils/helpers/nftMetadataDecoder.ts:5]: done

File: src/utils/services/Hip412Validator.ts (hunk 8 above)
Comment (michalrozekariane) [reply src/utils/services/Hip412Validator.ts:83]: done

File: src/utils/services/Hip412Validator.ts (hunk 11 above)
Comment (michalrozekariane) [reply src/utils/services/Hip412Validator.ts:116]: `file !== '.gitkeep'` deleted. Also I added a comment explaining what sorting is for.

File: src/utils/services/Hip412Validator.ts (hunk 10 above)
Comment (michalrozekariane) [reply src/utils/services/Hip412Validator.ts:57]: as we discussed - the error structure is a little bit different for array

File: src/utils/services/Hip412Validator.ts (hunk 9 above)
Comment (michalrozekariane) [reply src/utils/services/Hip412Validator.ts:79]: as we discussed - the error structure is a little bit different for array

File: src/utils/helpers/nftMetadataDecoder.ts (hunk 6 above)
Comment (michalrozekariane) [reply src/utils/helpers/nftMetadataDecoder.ts:11]: done

File: src/types/nfts.ts (hunk 5 above)
Comment (michalrozekariane) [reply src/types/nfts.ts:11]: `meta?` field deleted, it's not needed, good catch.

File: src/api/mirrorNode.ts (hunk 4 above)
Comment (michalrozekariane) [reply src/api/mirrorNode.ts:89]: done

File: src/api/mirrorNode.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+import axios from 'axios';
+import { NFTDetails, NFTS } from '../types/nfts';
+import { MetadataObject } from '../types/csv';
+import { dictionary } from '../utils/constants/dictionary';
+import { errorToMessage } from '../utils/helpers/errorToMessage';
+import { NetworkName } from '@hashgraph/sdk/lib/client/Client';
+
+const getMirrorNodeUrlForNetwork = (network: NetworkName): string => {
+  return `https://${network === 'mainnet' ? 'mainnet-public' : network}.mirrornode.hedera.com/api/v1`;
+};
```
[the remark this answers is not in the document]
Comment (michalrozekariane) [reply src/api/mirrorNode.ts:11]: I will, thanks.

**Review by se7enarianelabs — APPROVED** [review]

**Comment by BartoszSolkaArianelabs** [comment]
Merged in hedera-nft-utilities

**Pull request closed without merging**